sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -376,6 +387,23 @@ object ScalaReflection extends ScalaReflection {
           dataType = ObjectType(udt.getClass))
         Invoke(obj, "deserialize", ObjectType(udt.userClass), getPath :: Nil)
 
+      case t if isValueClass(t) =>
+        // nested value class is treated as its underlying type
+        // top level value class must be treated as a product
+        val underlyingType = getUnderlyingTypeOf(t)
+        val underlyingClsName = getClassNameFromType(underlyingType)
+        val clsName = t.typeSymbol.asClass.fullName
```

Comment:
how about `getClassNameFromType(t)`?

Reply:
I cannot use that because it returns the class name after erasure (e.g. it returns `Int` for `IntWrapper`). I'll create a separate method to make this clearer.